Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp

```diff
@@ -0,0 +1,818 @@
+ */
+
+#include "config.h"
+#include "DFGLoopUnrollingPhase.h"
+
+#if ENABLE(DFG_JIT)
+
+#include "CodeOrigin.h"
+#include "DFGBasicBlock.h"
+#include "DFGBlockInsertionSet.h"
+#include "DFGCFAPhase.h"
+#include "DFGGraph.h"
+#include "DFGNaturalLoops.h"
+#include "DFGNodeOrigin.h"
+#include "DFGNodeType.h"
+#include "DFGPhase.h"
+#include <wtf/HashSet.h>
+
+namespace JSC {
+namespace DFG {
+
+class LoopUnrollingPhase : public Phase {
+public:
+    static constexpr uint32_t MaxUnrolledLoopCount = 2;
+    static constexpr uint32_t MaxIterationCount = 4;
+    static constexpr uint32_t MaxLoopBodySize = 6;
+
+    using NaturalLoop = CPSNaturalLoop;
+
+    typedef bool (*ComparisonFunction)(CheckedInt32, CheckedInt32);
+    typedef CheckedInt32 (*UpdateFunction)(CheckedInt32, CheckedInt32);
+
+    struct LoopData {
+        BasicBlock* header() const { return loop->header().node(); }
+        Node* condition() const { return tail->terminal()->child1().node(); }
+        bool isInductionVariable(Node* node) { return node->operand() == inductionVariable->operand(); }
+        void dump(PrintStream& out) const;
+
+        const NaturalLoop* loop { nullptr };
+        BasicBlock* preHeader { nullptr };
+        BasicBlock* tail { nullptr };
+        BasicBlock* next { nullptr };
+
+        // for (i = initialValue; condition(i, operand); i = update(i, updateValue)) { ... }
+        Node* inductionVariable { nullptr };
+        CheckedInt32 initialValue { INT_MIN };
+        CheckedInt32 operand { INT_MIN };
+        Node* update { nullptr };
+        CheckedInt32 updateValue { INT_MIN };
+        CheckedUint32 iterationCount { 0 };
+    };
+
+    LoopUnrollingPhase(Graph& graph)
+        : Phase(graph, "Loop Unrolling"_s)
+        , m_blockInsertionSet(graph)
+    {
+    }
+
+    bool run()
+    {
+        bool changed = false;
+
+        if (UNLIKELY(Options::verboseLoopUnrolling())) {
+            dataLogLn("Graph before Loop Unrolling Phase:");
+            m_graph.dump();
+        }
+
+        uint32_t unrolledCount = 0;
+        while (true) {
+            const NaturalLoop* loop = selectDeepestNestedLoop();
+            if (!loop || unrolledCount >= MaxUnrolledLoopCount)
+                break;
+            if (!tryUnroll(loop))
+                break;
+
+            ++unrolledCount;
+            changed = true;
+        }
+
+        dataLogLnIf(Options::verboseLoopUnrolling(), "Successfully unrolled ", unrolledCount, " loops.");
+        return changed;
+    }
+
+    const NaturalLoop* selectDeepestNestedLoop()
+    {
+        m_graph.ensureCPSNaturalLoops();
+
+        const NaturalLoop* target = nullptr;
+        int32_t maxDepth = INT_MIN;
+
+        UncheckedKeyHashMap<const NaturalLoop*, int32_t> depthCache;
+        for (unsigned loopIndex = m_graph.m_cpsNaturalLoops->numLoops(); loopIndex--;) {
+            const NaturalLoop& loop = m_graph.m_cpsNaturalLoops->loop(loopIndex);
+
+            if (loop.size() > MaxLoopBodySize) {
```

Comment:
Maybe we should also check # of DFG nodes in this loop? Currently, it is checking # of basic blocks in this loop, and it is not directly corresponding to the size of body (# of DFG nodes).

Reply:
Added `maxLoopNodeSize` and move the check into `shouldUnrollLoop`.